src/java.net.http/share/classes/jdk/internal/net/http/LimitingSubscriber.java

```diff
@@ -0,0 +1,148 @@
+import jdk.internal.net.http.common.Utils;
+
+import java.io.IOException;
+import java.net.http.HttpResponse.BodySubscriber;
+import java.nio.Buffer;
+import java.nio.ByteBuffer;
+import java.util.List;
+import java.util.concurrent.CompletionStage;
+import java.util.concurrent.Flow.Subscription;
+
+import static java.util.Objects.requireNonNull;
+
+/**
+ * A subscriber limiting the maximum number of bytes that are allowed to be consumed by a downstream subscriber.
+ *
+ * @param <T> the response type
+ */
+public final class LimitingSubscriber<T> implements TrustedSubscriber<T> {
+
+    private final BodySubscriber<T> downstreamSubscriber;
+
+    private final long capacity;
+
+    private State state;
+
+    private long length;
+
+    private interface State {
+
+        enum Terminated implements State { INSTANCE }
+
+        record Subscribed(Subscription subscription) implements State {}
+
+    }
+
+    /**
+     * @param downstreamSubscriber the downstream subscriber to pass received data to
+     * @param capacity the maximum number of bytes that are allowed
+     * @throws IllegalArgumentException if {@code capacity < 0}
+     */
+    public LimitingSubscriber(BodySubscriber<T> downstreamSubscriber, long capacity) {
+        if (capacity < 0) {
+            throw new IllegalArgumentException("was expecting \"capacity >= 0\", found: " + capacity);
+        }
+        this.downstreamSubscriber = requireNonNull(downstreamSubscriber, "downstreamSubscriber");
+        this.capacity = capacity;
+    }
+
+    @Override
+    public void onSubscribe(Subscription subscription) {
+        requireNonNull(subscription, "subscription");
+        if (state != null) {
+            subscription.cancel();
+        } else {
+            state = new State.Subscribed(subscription);
+            downstreamSubscriber.onSubscribe(subscription);
+        }
+    }
+
+    @Override
+    public void onNext(List<ByteBuffer> buffers) {
+
+        // Check arguments
+        requireNonNull(buffers, "buffers");
+        assert Utils.hasRemaining(buffers);
+
+        // Short-circuit if not subscribed
+        if (!(state instanceof State.Subscribed subscribed)) {
+            return;
+        }
+
+        // See if we may consume the input
+        boolean lengthAllocated = allocateLength(buffers);
+        if (lengthAllocated) {
+            downstreamSubscriber.onNext(buffers);
+        }
+
+        // Otherwise, trigger failure
+        else {
+            state = State.Terminated.INSTANCE;
+            downstreamSubscriber.onError(new IOException(
+                    "the maximum number of bytes that are allowed to be consumed is exceeded"));
+            subscribed.subscription.cancel();
+        }
+
+    }
+
+    private boolean allocateLength(List<ByteBuffer> buffers) {
+        long bufferLength = buffers.stream().mapToLong(Buffer::remaining).sum();
+        long nextLength = Math.addExact(length, bufferLength);
```

Comment:
`Math.addExact` throws a `ArithmeticException` if there's an overflow during the addition. In its current form, this code can end up propagating the exception from here. Instead we should add a try/catch block to catch it and return false (implying the capacity has exceeded).

Reply:
Good catch!

Reply:
Fixed in 23f79a20476f784058801454a28d37f37df92982.